### ISSUE TITLE
Custom identifierRegexps is not working

### ISSUE THREAD

**Issue by rsmith31415** [ajaxorg/ace#2532]
This seems to be a recurrent issue, but maybe someone can give a bit of help figuring this out. 

The idea is to be able to use a custom `identifierRegex` in order to change the regex that is used in [retrievePrecedingIdentifier](https://github.com/ajaxorg/ace/blob/master/lib/ace/autocomplete/util.js#L50).

This should be possible providing an `identifierRegexps`  to the completers so that [this line takes effect: 

```
editor.completers.forEach(function(completer) {
        if (completer.identifierRegexps) {
            completer.identifierRegexps.forEach(function(identifierRegex) {
                if (!prefix && identifierRegex)
                    prefix = util.retrievePrecedingIdentifier(line, pos.column, identifierRegex);
            });
        }
```

**Comment by sugang** [comment]
I was able to reproduce your issue. For some reason the custom regexp doesn't work at all. I am researching a solution right now.

**Comment by rsmith31415** [comment]
Thanks. Let me know if you find a solution.

**Comment by sugang** [comment]
Ok. I figured out what was the problem. Some programmers were definitely drunk. 

Here's the solution to the problem: I was using the 'built' version. You need to change the `ext-language_tools.js`. My fix is not perfect, but it's much better than modifying the `ID_REGEX` thingy.

1) Look for the function `getCompletionPrefix`, as you mentioned. (around line 1900), make a copy. As you mentioned in your original post, this function is actually correct.

2) Look for the function this.gatherCompletions around line 1400. You will see these two lines:
```
    var prefix = util.retrievePrecedingIdentifier(line, pos.column); 
   ...
    prefix: util.retrievePrecedingIdentifier(line, pos.column, results[0] && results[0].identifierRegex),`
```
These are apparently wrong, as they are both not using the correct prefix which can be generated with `getCompletionPrefix` function. Since it's out of scope, I simply made a copy of this function, and replace the above two lines with 

```
    var prefix = this.getCompletionPrefix(editor);
   ...
    prefix: prefix
```

3) Add `identifierRegexps` to your custom completer. Note that the way ACE is doing autocompletion is it's looking at a single character at a time. Therefore you need to include all the possible permutations of the characters that could be part of the prefix all together. I added the hashtag.

4) Note if you add hashtag, hashtag then becomes part of the prefix. You need to append that to the results of your rhymeCompleter, otherwise the popup won't work.

```
$.getJSON(
	"http://rhymebrain.com/talk?function=getRhymes&word=" + prefix.substring(1),
	function(wordList) {
	// wordList like [{"word":"flow","freq":24,"score":300,"flags":"bc","syllables":"1"}]
		console.log(wordList);
		callback(null, wordList.map(function(ea) {
		return {name: '#' + ea.word, value: '#' + ea.word, score: ea.score, meta: "rhyme"}
	}));
```

I may actually rewrite this part of the library to add a 'trigger' symbol which won't be included as part of the term.

**Comment by rsmith31415** [comment]
I'm glad you figured it out. Do you have a branch to test your changes? Even better, you might want to create a pull request to see if this gets merged.

**Comment by sugang** [comment]
No.. I simply updated code on my local copy and it's working perfectly now. I think it's not a quite elegant fix as I made a duplicate function. I am still not very familiar with the overall structure of ACE, but I am curious why the getCompletionPrefix function was a local function which can't be referred by other pieces of ACE easily.

How to create a pull request?

**Comment by rsmith31415** [comment]
There are a lot of good answers [here](http://stackoverflow.com/questions/14680711/how-to-do-a-github-pull-request).
